src/librbd/api/Group.cc

```diff
@@ -438,6 +439,53 @@ int group_snap_rollback_by_record(librados::IoCtx& group_ioctx,
   return ret_code;
 }
 
+template <typename I>
+void notify_unquiesce(std::vector<I*> &ictxs,
+                      const std::vector<uint64_t> &requests) {
+  ceph_assert(requests.size() == ictxs.size());
+  int image_count = ictxs.size();
+  std::vector<C_SaferCond> on_finishes(image_count);
+
+  for (int i = 0; i < image_count; ++i) {
+    ImageCtx *ictx = ictxs[i];
+
+    ictx->image_watcher->notify_unquiesce(requests[i], &on_finishes[i]);
+  }
+
+  for (int i = 0; i < image_count; ++i) {
+    on_finishes[i].wait();
+  }
+}
+
+template <typename I>
+int notify_quiesce(std::vector<I*> &ictxs, ProgressContext &prog_ctx,
+                   std::vector<uint64_t> *requests) {
+  int image_count = ictxs.size();
+  std::vector<C_SaferCond> on_finishes(image_count);
+
+  requests->resize(image_count);
+  for (int i = 0; i < image_count; ++i) {
+    auto ictx = ictxs[i];
+
+    (*requests)[i] = ictx->image_watcher->notify_quiesce(prog_ctx,
+                                                         &on_finishes[i]);
+  }
+
+  int ret_code = 0;
+  for (int i = 0; i < image_count; ++i) {
+    int r = on_finishes[i].wait();
+    if (r < 0) {
```

Comment:
It will need to be solved at some point before Pacific, but we do need to support (optionally) ignoring `-EOPNOTSUPP` on quiesce calls or else we break backwards compatibility.

Reply:
As I see, `-EOPNOTSUPP` is returned by exclusive lock owner only, in the ack payload. The current implementation of quiesce notifications does not check the ack payload, so I believe (and I think I tested it when adding quiesce callbacks) we need nothing special to do here -- quiesce notifications will be just ignored by clients that do not support them and `group_snap_create` will proceed.

Reply:
That makes sense, thanks. It does make me think that perhaps we should add a return code to the `quiesce_complete` API so that we can return a failure. Right now, the only failure you can expect to get will be an `-ETIMEDOUT` (or maybe `-ENOENT` or `-EPERM` in some extreme cases). It looks like otherwise it would be encoding the actual quiesce return code in the ack payload via `C_ResponseMessage` -- but it's currently never decoded.

Reply:
Added this idea to my TODO. Not sure though, what users would like more, if `snap_create` failed when a callback failed or proceeded...

Reply:
... it just relates back to the user flags to control the quiesce behaviour. The current implementation behaviour is basically `ignore`/`best-effort` which is fine for most people I suspect.